contracts/ConfidentialERC20/ConfidentialERC20.sol

```diff
@@ -260,14 +276,18 @@ abstract contract ConfidentialERC20 is Ownable, IConfidentialERC20, IERC20Metada
         burnRqs[requestID] = BurnRq(account, amount);
     }
 
-    function _burnCallback(uint256 requestID, bool decryptedInput) public virtual onlyGateway {
+    error InsufficientBalanceToBurn(address account, uint64 burnAmount);
+
+    function _burnCallback(uint256 requestID, bool hasEnoughBalance) public virtual onlyGateway {
         BurnRq memory burnRequest = burnRqs[requestID];
         address account = burnRequest.account;
         uint64 amount = burnRequest.amount;
-        if (!decryptedInput) {
-            revert("Decryption failed");
+        if (!hasEnoughBalance) {
+            revert InsufficientBalanceToBurn(account, amount);
         }
         _totalSupply = _totalSupply - amount;
+        // Unlock the burn amount
+        _lockedBalances[account] = TFHE.sub(_lockedBalances[account], amount);
```

Comment:
This is underflow free because we know we have locked up amount before this callback may run